### PR TITLE
[ARRCONN-31] Fixed search by date execution query

### PR DESCRIPTION
**Pull request by vladimir-trifonov**
https://jira.appcelerator.org/browse/ARRCONN-31

### CODE REVIEW

File: lib/utility/convertJSDateToJSForceDate.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ * Converts JS Date Objects to JS Force Date Literals.
+ * @param {*}
+ * @returns object
+ */
+exports.convertJSDateToJSForceDate = function convertJSDateToJSForceDate(obj) {
+	if (obj instanceof Date) {
+		return jsforce.Date.toDateLiteral(obj);
+	} else if (typeof(obj) === 'string') {
+		return jsforce.Date.toDateLiteral(Date.parse(obj));
+	} else if (typeof(obj) === 'object') {
```
Comment (janvennemann) [inline lib/utility/convertJSDateToJSForceDate.js]: What's the use case for this special object type handling?

Comment (vladimir-trifonov) [inline lib/utility/convertJSDateToJSForceDate.js]: Hi Jan,
Do you mean this condition here 

`else if (typeof(obj) === 'string') {`

If so, it's just for precaution. I think it won't be necessary.

About `else if (typeof(obj) === 'object') {` it's doing recursive data converting.

Comment (janvennemann) [inline lib/utility/convertJSDateToJSForceDate.js]: Ahhh yes, of course! Thank's for the clarification ;)

**Comment by vladimir-trifonov** [comment]
@janvennemann Should be fine, right now.

File: lib/methods/query.js [hunk 2]
```diff
@@ -19,8 +20,29 @@ exports.query = function (Model, options, callback) {
...
+		 * Convert Date filter props to SalesForce based date fields
+		 */
+		if (options.where) {
+			Object.keys(options.where).filter(function (fieldKey) {
+				return _.has(Model.fields, fieldKey) && ['date', 'time', 'datetime'].indexOf(Model.fields[fieldKey].originaltype) !== -1;
+			}).forEach(function (fieldKey) {
+				// Nested iteration
+				var convert = function convert(obj, type) {
+					if (typeof(obj) === 'object') {
+						Object.keys(obj).map(function (key) {
```
Comment (janvennemann) [inline lib/methods/query.js]: Consider using `forEach` if you do not use the return value of `map`.

File: lib/generator.js [hunk 3]
```diff
@@ -36,6 +36,7 @@ exports.generateModels = function(connectorName, modelDir, schema, autogen, cb)
 			var newField = model.fields[field.name] = {
 				name: field.name,
 				type: translateFieldToType(field, objectName),
+				originaltype: field.type,
```
Comment (janvennemann) [inline lib/generator.js]: Please use lower camel case for variable names: `originalType`

File: lib/utility/convertDateToJSFDateTime.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+var jsforce = require('jsforce');
+
+/**
+ * Convert Date filter props to SalesForce based date fields
+ * @param {field} value
+ * @param {fieldType} salesforce type value
```
Comment (janvennemann) [inline lib/utility/convertDateToJSFDateTime.js]: The correct order for `@param` JSDoc annotation is type, name and optional description. Please change to something like this: `@param {string} fieldType Original type from salesforce schema`. Same applies to the first parameter.

File: lib/utility/convertDateToJSFDateTime.js [hunk 5]
```diff
@@ -2,8 +2,8 @@ var jsforce = require('jsforce');
 
 /**
  * Convert Date filter props to SalesForce based date fields
- * @param {field} value
- * @param {fieldType} salesforce type value
+ * @param {field} Date in a string format
+ * @param {fieldType} Original type from salesforce schema
```
Comment (janvennemann) [inline lib/utility/convertDateToJSFDateTime.js]: This is still wrong. The curly braces should contain the type, not the param name. So this should be `@param {string} fieldType Original type from salesforce schema`. Same with the first param. See http://usejsdoc.org/tags-param.html 😉

**Comment by vladimir-trifonov** [comment]
yes, you're right. it's done

**Comment by janvennemann** [comment]
Code looks gode now! One last thing though. Can you reword you commits so they reference the correct ticket number? The project key in your commits should be `ARRCONN`. Thanks!

**Comment by vladimir-trifonov** [comment]
Done! Look into [https://github.com/appcelerator/appc.salesforce/pull/5](https://github.com/appcelerator/appc.salesforce/pull/5).

**Comment by janvennemann** [comment]
Closing in favor of #5

**Pull request closed without merging**